resource/geo-density-chart.js

```diff
@@ -0,0 +1,379 @@
+            .on('mouseover', function() { // handle the event user mouse is over the circle data point
+                _tooltip.style("visibility", "visible")
+                    .style("left", (d3.event.pageX+15) + "px")
+                    .style("top", (d3.event.pageY-25) + "px");
+                d3.select(this).transition()
+                    .duration('50')
+                    .attr("fill", "red")
+                    .attr('r', 10)
+            })
+            .on('mouseout', function() { // handle
+                _tooltip.style("visibility", "hidden");
+                d3.select(this).transition()
+                    .duration('150')
+                    .attr("r", 5)
+            })
+            // potentially useless
+            .attr("r", 5)
+            .attr("stroke", "black")
+            .attr("stroke-width", 3)
+            .attr("fill-opacity", .4)
+            .attr("fill", "transparent")
+    };
+
+    this.addUpdateEvent = function(){
+        _mapPointer.on('moveend', function(){
+            d3.selectAll("circle")
+                .attr("cx", function(d){ return _mapPointer.latLngToLayerPoint([d.lat, d.long]).x })
+                .attr("cy", function(d){ return _mapPointer.latLngToLayerPoint([d.lat, d.long]).y })
+        })
+    };
+
+    /**
+     * Render the heatmap. Take into account pre-stored geospatial data and visual layout parameters.
+     */
+    this.render = function(){
+        _quadtree = d3.quadtree()
+            .extent([[-100, -100], [100, 100]])
+            .addAll(_coreData);
+        if (_initialised){
+            console.log("updateMapVisualisation: Remove existing layer");
+            _heatMapLayer.remove();
+        }
+        let weightedGeopoints = [];
+        for (let i = 0; i < _coreData.length; i++) {
+            weightedGeopoints.push([_coreData[i][0], _coreData[i][1], _coreData[i][2]*_intensity]);
+        }
+        _heatMapLayer = L.heatLayer(weightedGeopoints, {radius: _radius, gradient: _gradient});
+        // Add layer to global object main map (leaflet target map)
+        _heatMapLayer.addTo(_mapPointer);
+        _initialised = true;
+    };
+
+    /**
+     * Clear heatmap layer.
+     */
+    this.clear = function(){
+        _coreData = [];
+        _quadtree = null;
+        // Clear a leaflet layer if it has been initialised
+        if (_initialised){
+            _heatMapLayer.remove();
+        }
+    }
+}
+
+
+/**
+ * Format the tooltip to be displayed in HTML
+ * tooltip = {"reviews_per_month": 0.92, "room_type": "Private Room"};
+ * @param tooltip
+ * @returns {string}
+ */
+function formatTooltip(tooltip){
+    if (tooltip && Object.keys(tooltip).length === 0 && tooltip.constructor === Object) {
+        return ""
+    }
+    var formattedString = "<hr>";
+    for (var key of Object.keys(tooltip)) {
+        formattedString += key+": ";
+        formattedString += "<b>"+tooltip[key]+"</b>";
+        formattedString += "<br>";
+    }
```

Comment:
Better to use this syntax
```suggestion
    let formattedString = "<hr>";
    for (const [key, value]  of Object.entries(tooltip)) {
        formattedString += `${key}: <b>${value}</b><br>`
    }
```

Reply:
Done in c71463d